### PR TITLE
fix: make local MCP safe for caller-owned ready indexes

### PR DESCRIPTION
**Pull request by zcl0621**
## Problem

Local stdio MCP already holds a live `SharedStores`, but a shared vector-store read failure could fall through to `VectorStore::new` and attempt to open the same LMDB environment again. That replaced the useful search failure with `LMDB double-open prevented`.

Callers that prepare an index before starting MCP also had no explicit way to prevent background mutation or reject an incomplete index before the stdio transport opened. `--create-index=false` controls creation only; overloading it with either behavior would change its existing semantics.

## Fix

- Disable the standalone vector-store reopen fallback whenever a live shared store exists; preserve the original failure in the returned error chain.
- Add independent local-MCP controls:
  - `--readonly` opens the vector and full-text stores read-only and skips incremental refresh and file watching.
  - `--require-ready` fails startup unless the live stores contain chunks, a built vector index, full-text documents, and metadata explicitly marks the index non-partial.
- Reject both flags in `auto` and `client` modes, where they cannot control the remote serve process.
- Prevent either flag from creating a missing index while leaving existing `--create-index` behavior unchanged for writable local MCP.
- Preserve the existing `run_mcp_server` API and expose the additional controls through `run_mcp_server_with_options`.

## Verification

- `cargo fmt`
- `cargo clippy --all-targets -- -D warnings`
- `cargo test --lib -- --nocapture`: 627 passed, 0 failed, 13 ignored
- `cargo test --test '*' -- --nocapture --skip ignore`: 18 passed, 0 failed, 4 ignored
- Runtime regression tests cover the live-`SharedStores` LMDB path and readiness checks against populated vector/FTS stores plus complete, partial, missing-marker, and malformed metadata.
- Final review found no High/Critical issues; the two Medium test gaps were fixed before opening this PR.

`CHANGELOG.md` is updated under the pending 1.3.4 section.

### CODE REVIEW

File: src/mcp/mod.rs [hunk 1]
```diff
@@ -2260,38 +2329,27 @@ impl CodesearchService {
...
                     tracing::error!("Shared vector store read failed: {:?}", shared_err);
 
-                    // In serve mode, do NOT fall back to a standalone VectorStore —
-                    // it would open a second LMDB handle on the same .codesearch.db,
-                    // which LMDB rejects ("environment already opened with different options").
-                    if self.serve_state.is_some() {
+                    // Live SharedStores already owns the LMDB env (stdio and serve).
+                    // Do not open a second VectorStore on the same db_path.
+                    if !allow_vector_store_second_open(self.shared_stores.is_some()) {
                         return Err(shared_err.context(
```
[the remark this answers is not in the document]
Comment (zcl0621) [reply src/mcp/mod.rs:1204]: Agreed — this arm already has a live SharedStores, so `is_some()` was a tautology. Follow-up commit now passes `true` here so the gate is explicit.

**Comment by flupkede** [comment]
Maintainer adaptation for the upcoming `src/mcp/mod.rs` split (#223) — your two commits are preserved verbatim as the feature commits (authorship intact); one adaptation commit was added on top:

- **Rebased onto `refactor/mcp-mod-split`** and ported `run_mcp_server_with_options`, `McpStartupOptions`, `validate_local_startup_flags`, `may_create_missing_index`, `validate_prebuilt_index_health` and `require_prebuilt_index_ready` to `mcp/runtime.rs`. They are re-exported via `pub use runtime::*`, so the `crate::mcp::...` call sites (including `src/cli/mod.rs`) are unchanged. `allow_vector_store_second_open` stays in `mcp/mod.rs` next to `with_vector_store_read_for`.
- **Removed the brittle tests** flagged in review: the three `include_str!("mod.rs")` source-scanning tests (they would have broken mechanically under the module split) and the duplicate serve-gate test. The runtime tests remain as the behavioral contract: LMDB error preservation, readiness edge cases (complete/partial/missing/malformed metadata), and flag validation.
- Helpers are `pub(crate)` so `mcp::tests` can reach them across the module boundary.

Validation: `cargo fmt`, `cargo check --all-targets`, `cargo clippy --all-targets -- -D warnings`, 148 `mcp::tests` + 6 CLI flag tests, all green.

Base retargeted to `refactor/mcp-mod-split` so this stacks on #223; GitHub will re-target to `develop` automatically once that merges. @zcl0621 no action needed from you unless you object to the test removals.

**Comment by zcl0621** [comment]
No objection to the test removals — thanks for the adaptation, makes sense given the mcp module split. Happy to merge once #223 lands.

**Review by flupkede — APPROVED** [review]
Reviewed in full (see adaptation commit 88580a4): core LMDB second-open fix verified correct against develop (index_health, FTS num_documents, SharedStores::new_readonly, partial marker writer all checked); readiness gate verified against indexer behaviour (partial:false on success, partial:true on cancel, fail-safe on missing marker). The brittle include_str! source-scanning tests and the duplicate serve-gate test were removed in the port commit; runtime tests remain the behavioural contract. Validated: cargo fmt, check --all-targets, clippy -D warnings, 148 mcp + 6 cli tests green. Approving on behalf of the maintainers — thanks for the clean fix!
